File: leetcode/stack/155. Min Stack/cpp/main.cpp
class MinStack {
    private:
        stack<int> st;
        stack<pair<int, int>> minStack;
public:
    /** initialize your data structure here. */
    MinStack() {}
    
    void push(int x) {
        if (minStack.size() == 0 || minStack.top().first > x) {
            minStack.push({ x, 1 });
        } else if (minStack.top().first == x) {
            pair<int, int> p = minStack.top();
            minStack.pop();
            minStack.push({ x, p.second + 1 });
        }
        
        st.push(x);
    }
    
    void pop() {
        int x = st.top();
        st.pop();
        
        if (minStack.top().first == x) {
            pair<int, int> p = minStack.top();
            minStack.pop();
            minStack.push({ x, p.second - 1 });
            
            if (minStack.top().second <= 0) {
                minStack.pop();
            }
        }
    }
    
    int top() {
        return st.top();
    }
    
    int getMin() {
        return minStack.top().first;
    }
};

File: leetcode/stack/155. Min Stack/py/main.py
import collections

class MinStack:

    def __init__(self):
        self.stack = []
        self.min_backet = []


    def push(self, x: int) -> None:
      self.stack.append(x)

      if not self.min_backet or self.min_backet[-1] >= x:
        self.min_backet.append(x)
      

    def pop(self) -> None:
      x = self.stack.pop()

      if self.min_backet[-1] == x:
        self.min_backet.pop()
      

    def top(self) -> int:
      return self.stack[-1]
      
      
    def getMin(self) -> int:
        return self.min_backet[-1]

